File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:   
    def prune(self, heads):
        pruneList = []
        for i in range(len(heads)):
            if heads[i] is None:
                pruneList.append(i)
        offset = 0
        for ind in pruneList:
            heads.pop(ind + offset)
            offset -= 1
        return heads


    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        newHead = None
        headToUpdate = None
        # loop through all heads, find the minimal one and set newHead to that one
        smallest = 1001
        for i in range(len(lists)):
            head = lists[i]
            if head is None:
                continue
            if head.val <= smallest:
                newHead = head
                smallest = head.val
                headToUpdate = i
        # return newHead
        if not headToUpdate is None:
            if lists[headToUpdate].next:
                lists[headToUpdate] = lists[headToUpdate].next
            else:
            # remove from the list
                lists.pop(headToUpdate)   
        headToReturn = newHead
        # I have the head of my merged list, and a list of pointers to all of the other heads
        # while there are valid pointers, I need to get the 
        iters = 0
        while True:
            # prune the elements of the list that are None
            lists = self.prune(lists)
            if len(lists) == 0:
                return headToReturn
            mini = 1001
            headToUpdate = None
            for i in range(len(lists)):
                head = lists[i]
                if head.val <= mini:
                    mini = head.val
                    headToUpdate = i
            newHead.next = lists[headToUpdate]
            lists[headToUpdate] = lists[headToUpdate].next
            newHead = newHead.next # move forward
        
            

File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == nullptr) {
            return nullptr;
        }
        ListNode* looper = head;
        ListNode* dummy = nullptr; 
        ListNode* scout;
        while (looper && looper->next != nullptr) {
            std::cout << looper->val << std::endl;
            scout = looper->next;
            looper->next = dummy;
            dummy = looper;
            looper = scout;
        }
        looper->next = dummy;

        std::cout << head->val << std::endl;
        if (head->next) {
             std::cout << head->next->val << std::endl;
        }
        return looper;
    }
};
